### PR TITLE
Tidy up the NativeValueIndex

### PR DESCRIPTION
**Pull request by adamretter**
Code audit and cleanup in the NativeValueIndex

### CODE REVIEW

File: src/org/exist/storage/NativeValueIndex.java [hunk 1]
```diff
@@ -851,6 +852,7 @@ public NodeSet matchAll(final XQueryWatchDog watchDog, final DocumentSet docs, f
             final int collectionId = iter.next().getId();
 
             watchDog.proceed(null);
+            watchDog.proceed(null);
```
Comment (adamretter) [inline src/org/exist/storage/NativeValueIndex.java]: @shabanovd Oops! No idea how that happened, I just fixed, rebased and pushed again :-)

**Comment by shabanovd** [comment]
@adamretter test fail

**Comment by adamretter** [comment]
@shabanovd Hopefully should be good now... I had some uncommitted changes locally.

File: src/org/exist/storage/io/AbstractVariableByteInput.java [hunk 2]
```diff
@@ -142,36 +143,35 @@ public int read(byte b[], int off, int len) throws IOException {
...
-    
-    public void copyRaw(VariableByteOutputStream os, int count) throws IOException {
-        for (int i = 0; i < count; i++) {
-            os.buf.append((byte) read());
-        }
+
+    @Override
+    public void copyRaw(final VariableByteOutputStream os, final int count) throws IOException {
+        final byte buf[] = new byte[count];
+        read(buf, 0, count);
```
Comment (shabanovd) [inline src/org/exist/storage/io/AbstractVariableByteInput.java]: `read` can read less bytes, so `write` must be called with len returned by `read`, no?

File: src/org/exist/xquery/GeneralComparison.java [hunk 3]
```diff
@@ -377,7 +377,7 @@ else if( key.getType() != indexType ) {
                 final Collator collator   = ( ( collationArg != null ) ? getCollator( contextSequence ) : null );
 
                 if( truncation == Constants.TRUNC_NONE ) {
-                    temp         = context.getBroker().getValueIndex().find(context.getWatchDog(), relation, contextSequence.getDocumentSet(), contextSet, NodeSet.DESCENDANT, contextQName, ( Indexable )key, collator );
+                    temp         = context.getBroker().getValueIndex().find(context.getWatchDog(), relation, contextSequence.getDocumentSet(), contextSet, NodeSet.DESCENDANT, contextQName, ( Indexable )key);
```
Comment (shabanovd) [inline src/org/exist/xquery/GeneralComparison.java]: why do you remove `collator` here?

File: src/org/exist/storage/NativeValueIndex.java [hunk 4]
```diff
@@ -327,211 +335,198 @@ public void sync() {
     @Override
     public void flush() {
         //TODO : return if doc == null? -pb
```
Comment (shabanovd) [inline src/org/exist/storage/NativeValueIndex.java]: remove or fix TODO

File: src/org/exist/storage/NativeValueIndex.java [hunk 5]
```diff
@@ -327,211 +335,198 @@ public void sync() {
...
+    private <T> void flush(final PendingChanges<T> pending, final FunctionE<T, Value, EXistException> dbKeyFn) {
+        for (final Map.Entry<T, List<NodeId>> entry : pending.changes.entrySet()) {
+            final T key = entry.getKey();
 
-                //Don't forget this one
-                FastQSort.sort(gids, 0, gidsCount - 1);
-                os.clear();
-                os.writeInt(this.doc.getDocId());
-                os.writeInt(gidsCount);
+            //TODO : NativeElementIndex uses ArrayLists -pb
```
Comment (shabanovd) [inline src/org/exist/storage/NativeValueIndex.java]: remove or fix TODO

File: src/org/exist/xquery/GeneralComparison.java (hunk 3 above)
Comment (adamretter) [inline src/org/exist/xquery/GeneralComparison.java]: @shabanovd Because it was unused.

**Comment by adamretter** [comment]
@shabanovd Okay I made those changes you suggested

File: src/org/exist/storage/NativeValueIndex.java [hunk 6]
```diff
@@ -324,211 +332,196 @@ public void sync() {
 
     @Override
     public void flush() {
-        //TODO : return if doc == null? -pb
-        final int keyCount = pending[IDX_GENERIC].size() + pending[IDX_QNAME].size();
-
-        if (keyCount == 0) {
+        if (doc == null || (pendingGeneric.changes.isEmpty() && pendingQName.changes.isEmpty())) {
```
Comment (shabanovd) [inline src/org/exist/storage/NativeValueIndex.java]: how are you sure that `doc == null` must affect `flush`?

Comment (shabanovd) [inline src/org/exist/storage/NativeValueIndex.java]: ok, it affect because of `doc.getCollection()`

Comment (adamretter) [inline src/org/exist/storage/NativeValueIndex.java]: @shabanovd exactly!

**Comment by adamretter** [comment]
@shabanovd Thanks. Together we got there :-)

**Comment by shabanovd** [comment]
@adamretter now time for https://github.com/eXist-db/exist/pull/1027 ? it also block my next two PRs.

**Comment by caseydawsonjordan** [comment]
Nice work guys!

**Comment by wolfgangmm** [comment]
This seems to have killed id() lookups. Try to store

```xml
<test xml:id="abc"/>
```

and query it with:

```xquery
xquery version "3.0";

doc("/db/test.xml")/id("abc")
```

**Comment by shabanovd** [comment]
it'll be good to add this as test

**Comment by adamretter** [comment]
@wolfgangmm Are you sure? That seems to work just fine for me here. Although I note that the following never worked in eXist (but should):

```xquery
document {
    <test xml:id="abc"/>
}/id("abc")
```

**Comment by wolfgangmm** [comment]
@adamretter Yeah, I'm not sure why storing test.xml and looking up the id fails. We do have tests in XPathQueryTest.ids() and they pass. Still all apps which rely on fn:id somewhere suddenly stopped working for me. I first thought it's the experimental code I was using, but I got the same on a clean clone. 

I'll report if I find out more.

**Comment by adamretter** [comment]
see also - https://github.com/eXist-db/exist/pull/1037

**Comment by adamretter** [comment]
@wolfgangmm if I store `test.xml` and run the XQuery you have above, then it works correctly. I am not sure why you think it fails? I think it also works for @shabanovd

**Pull request merged**